### PR TITLE
Initial KNN classifier add.

### PR DESCRIPTION
**Pull request by chukarsten**
### Pull Request Description
First crack at adding KNN classifier(/regressor) to the baseline.

-----
*After creating the pull request: in order to pass the **release_notes_updated** check you will need to update the "Future Release" section of* `docs/source/release_notes.rst` *to include this pull request by adding :pr:`123`.*

### CODE REVIEW

File: evalml/tests/automl_tests/test_automl.py
Comment (freddyaboulton) [inline evalml/tests/automl_tests/test_automl.py:1863]: How come splitting this test into many tests fixes the crashing worker problem?

File: evalml/pipelines/components/estimators/classifiers/kneighbors_classifier.py
Comment (freddyaboulton) [inline evalml/pipelines/components/estimators/classifiers/kneighbors_classifier.py:23]: I think this estimator would work for time series problems?

File: evalml/tests/model_understanding_tests/prediction_explanations_tests/test_algorithms.py
Comment (freddyaboulton) [inline evalml/tests/model_understanding_tests/prediction_explanations_tests/test_algorithms.py:111]: How come we are not allowing KNN for multiclass problems in shap?

File: evalml/pipelines/components/estimators/classifiers/kneighbors_classifier.py
Comment (freddyaboulton) [inline evalml/pipelines/components/estimators/classifiers/kneighbors_classifier.py:16]: I wonder if there's a different way to parametrize this? I imagine that for some datasets, the estimator might need more than 12 neighbors to get good performance.

Maybe we can parametrize this as the percentage of the training data used in the neighborhood. This is not blocking merge, I'm just thinking out loud hehe.

**Review by freddyaboulton — APPROVED** [review]
@chukarsten This looks good to me! I want to resolve the discussion on prediction explanations before merging though. 

Also, we typically exclude estimators from AutoMLSearch until we run the performance tests on them. We do this by adding the class name to `_not_used_in_automl` in `gen_utils.py`. I think we should follow the same pattern here but what do you think @dsherry ?

File: docs/source/release_notes.rst
Comment (bchen1116) [inline docs/source/release_notes.rst:13]: I think it'd be better to say `Add KNN Classifier to estimators` rather than to baseline?

File: evalml/pipelines/components/estimators/classifiers/kneighbors_classifier.py
Comment (bchen1116) [inline evalml/pipelines/components/estimators/classifiers/kneighbors_classifier.py:14]: nit-pick: Can we change the name to simply be `KNN Classifier`? Thinking from the user perspective, it might be annoying to type `K-Nearest Neighbors Classifier` if they're defining a component graph or parameters, especially since none of our other components has a dash in it

File: evalml/pipelines/components/estimators/classifiers/kneighbors_classifier.py
Comment (bchen1116) [inline evalml/pipelines/components/estimators/classifiers/kneighbors_classifier.py:41]: Is there a reason we aren't passing in `random_state` to this `super().__init__`?

File: evalml/pipelines/components/estimators/classifiers/kneighbors_classifier.py
Comment (bchen1116) [inline evalml/pipelines/components/estimators/classifiers/kneighbors_classifier.py:50]: Are we just saying that all features have 0 importance since it's a Nearest Neighbors alg? Just double checking

File: evalml/tests/automl_tests/test_automl.py
Comment (bchen1116) [reply evalml/tests/automl_tests/test_automl.py:1863]: The dreaded windows test bug D:

**Review by bchen1116 — COMMENTED** [review]
Left a few comments. Don't forget to add the KNN classifier to the `api_references.rst` file and make sure the corresponding docs are updated to include the classifier. 

Otherwise, agree with @freddyaboulton that this should be excluded from AutoMLSearch until perf tests are run.

File: evalml/pipelines/components/estimators/classifiers/kneighbors_classifier.py
Comment (angela97lin) [inline evalml/pipelines/components/estimators/classifiers/kneighbors_classifier.py:47]: Is this an internal note or documentation for the users 👀

File: evalml/tests/automl_tests/test_automl.py
Comment (angela97lin) [inline evalml/tests/automl_tests/test_automl.py:1880]: Comment no longer necessary? :D

**Review by angela97lin — APPROVED** [review]
Agreed with what's already been said about perf testing and adding to AutoML, but otherwise just left some comments for cleanup! 😁

File: evalml/tests/automl_tests/test_automl.py
Comment (chukarsten) [reply evalml/tests/automl_tests/test_automl.py:1863]: That I'm not sure.  We can't say conclusively that it does.  It fixed it "this time", but as @bchen1116 knows, sometimes just modifying something seems to affect the presence of the crashing worker.  We were discussing maybe having a nightly testing of main to see if main, outside of tests via merging, exhibits this crash going forward.  But really I guess this will be a question of whether we accept the additional lines due to splitting the test in the hopes of fixing the bug going forward, or not.

File: evalml/pipelines/components/estimators/classifiers/kneighbors_classifier.py
Comment (chukarsten) [reply evalml/pipelines/components/estimators/classifiers/kneighbors_classifier.py:23]: Got it, added.

File: docs/source/release_notes.rst
Comment (chukarsten) [reply docs/source/release_notes.rst:13]: Yup, thank you sir

File: evalml/pipelines/components/estimators/classifiers/kneighbors_classifier.py
Comment (chukarsten) [reply evalml/pipelines/components/estimators/classifiers/kneighbors_classifier.py:14]: Can do!

File: docs/source/release_notes.rst
Comment (chukarsten) [reply docs/source/release_notes.rst:13]: It's almost like I forgot to re-read the contrib guide ;)

File: evalml/pipelines/components/estimators/classifiers/kneighbors_classifier.py
Comment (chukarsten) [reply evalml/pipelines/components/estimators/classifiers/kneighbors_classifier.py:10]: Here I just tried to use Sklearn's naming like we do in the other estimators

File: evalml/pipelines/components/estimators/classifiers/kneighbors_classifier.py
Comment (chukarsten) [reply evalml/pipelines/components/estimators/classifiers/kneighbors_classifier.py:42]: I don't know why I made this decision - I've changed it back to mirror the other classes.

File: evalml/pipelines/components/estimators/classifiers/kneighbors_classifier.py
Comment (chukarsten) [reply evalml/pipelines/components/estimators/classifiers/kneighbors_classifier.py:52]: Yea, I'm using it as a substitute/placeholder for another algorithm.  We can't do feature_importance out of the box and we'll have to agree upon a technique to use.  I didn't want to change the interface by adding args to feature_importance and I tried raising a NotImplementedError, but that broke too much.

File: evalml/pipelines/components/estimators/classifiers/kneighbors_classifier.py
Comment (chukarsten) [reply evalml/pipelines/components/estimators/classifiers/kneighbors_classifier.py:49]: @angela97lin It was, but I updated the base class doc to reflect this.  I am pretty sure the Estimator base is comment rot.

File: evalml/tests/model_understanding_tests/prediction_explanations_tests/test_algorithms.py
Comment (chukarsten) [reply evalml/tests/model_understanding_tests/prediction_explanations_tests/test_algorithms.py]: It's been so long I don't even remember why I took this out.  I fixed it!

File: evalml/tests/model_understanding_tests/prediction_explanations_tests/test_algorithms.py
Comment (bchen1116) [inline evalml/tests/model_understanding_tests/prediction_explanations_tests/test_algorithms.py:109]: Is there a reason why we split this check?

**Review by bchen1116 — APPROVED** [review]
Left one note, but looks good to me! 

One thing I want to bring to attention would be the possibility of standardizing the names. I brought up the idea to change the string name to `KNN Classifier`, but I see the value of keeping the naming in line with how SKLearn names it, which was `KNeighborsClassifier`. I think we should make these names the same, like we do for all other components (ie class `LogisticRegressionClassifier` with name `Logistic Regression Classifier`). On this note, I think it might make sense to change the name to `K Neighbors Classifier` (no dash), but let me know what you think!

File: evalml/tests/model_understanding_tests/prediction_explanations_tests/test_algorithms.py
Comment (chukarsten) [reply evalml/tests/model_understanding_tests/prediction_explanations_tests/test_algorithms.py:109]: Sorry, that's right.  I split it when I was still skipping for KNN.  I'll put it back!

**Pull request merged**
